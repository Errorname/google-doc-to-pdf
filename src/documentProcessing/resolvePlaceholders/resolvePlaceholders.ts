import { Placeholder, Formatters } from '../types'

import computeContent from './computeContent'

const resolvePlaceholders = (
  placeholders: Placeholder[],
  data: Record<string, any>,
  options?: {
    formatters?: Formatters
    strict?: boolean
  }
): Placeholder[] => {
  return placeholders.map(placeholder => {
    const rawWithoutBrackets = placeholder.raw.slice(2, -2).trim()

    // CommentPlaceholder
    if (rawWithoutBrackets.startsWith('#')) {
      return {
        ...placeholder,
        type: 'comment',
        value: rawWithoutBrackets.slice(1).trim()
      }
    }

    if (['%', '@'].includes(rawWithoutBrackets[0])) {
      // Reserved for later usage
      return placeholder
    }

    // ContentPlaceholder
    return computeContent(placeholder, data, options)
  })
}

export default resolvePlaceholders
