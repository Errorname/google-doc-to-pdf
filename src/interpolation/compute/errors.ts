export class UnknownFormatterError extends Error {}

export class UnvalidFormatterError extends Error {}

export class UndefinedVariableError extends Error {}
