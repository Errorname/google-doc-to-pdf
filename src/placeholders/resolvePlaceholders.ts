import { Formatters, Placeholder, TypedPlaceholder } from '../types'
import { compute, parser } from './syntax'

const resolvePlaceholders = (
  placeholders: Placeholder[],
  data: Record<string, unknown>,
  options: {
    formatters: Formatters
    strict: boolean
  }
): TypedPlaceholder[] => {
  return placeholders.map((placeholder) => {
    const rawWithoutBrackets = placeholder.raw.slice(2, -2).trim()

    // Comment Placeholder
    if (rawWithoutBrackets.startsWith('#')) {
      return {
        ...placeholder,
        type: 'comment',
        value: rawWithoutBrackets.slice(1).trim(),
      }
    }

    // Future Placeholders
    if (['%', '@'].includes(rawWithoutBrackets[0])) {
      // Reserved for later usage
      return {
        ...placeholder,
        type: 'unknown',
      }
    }

    // Content Placeholder
    const expression = parser(rawWithoutBrackets)
    const result = compute(expression, data, options)
    return {
      ...placeholder,
      type: 'content',
      value: result,
    }
  })
}

export default resolvePlaceholders
