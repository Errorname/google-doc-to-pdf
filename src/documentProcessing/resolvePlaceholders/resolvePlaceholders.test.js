import resolvePlaceholders from './resolvePlaceholders'
import { UndefinedVariableError, UnknownFormatterError } from './errors'

describe('resolvePlaceholders', () => {
  test('no placeholders', () => {
    const placeholders = resolvePlaceholders([], {})

    expect(placeholders).toEqual([])
  })

  test('contentPlaceholder - text', () => {
    const placeholders = [
      {
        raw: '{{ user.name.first | lowercase }}',
        position: {
          start: 134,
          end: 166
        }
      }
    ]

    const contentPlaceholders = resolvePlaceholders(placeholders, {
      user: { name: { first: 'Thibaud' } }
    })

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('contentPlaceholder - image', () => {
    const placeholders = [
      {
        raw: '{{ catUrl | image }}',
        position: {
          start: 134,
          end: 166
        }
      }
    ]

    const contentPlaceholders = resolvePlaceholders(placeholders, {
      catUrl: 'http://example.cat'
    })

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('contentPlaceholder - image (width)', () => {
    const placeholders = [
      {
        raw: '{{ catUrl | image(150) }}',
        position: {
          start: 134,
          end: 166
        }
      }
    ]

    const contentPlaceholders = resolvePlaceholders(placeholders, {
      catUrl: 'http://example.cat'
    })

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('contentPlaceholder - image (with, height)', () => {
    const placeholders = [
      {
        raw: '{{ catUrl | image(150, 80) }}',
        position: {
          start: 134,
          end: 166
        }
      }
    ]

    const contentPlaceholders = resolvePlaceholders(placeholders, {
      catUrl: 'http://example.cat'
    })

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('commentPlaceholder', () => {
    const placeholders = [
      {
        raw: '{{ # This is a comment }}',
        position: {
          start: 134,
          end: 158
        }
      }
    ]

    const contentPlaceholders = resolvePlaceholders(placeholders, {})

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('unknownPlaceholder - reserved symbol', () => {
    const placeholders = [
      {
        raw: '{{ % reserved symbol }}',
        position: {
          start: 134,
          end: 158
        }
      }
    ]

    const contentPlaceholders = resolvePlaceholders(placeholders, {})

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('Unknown Formatter — strict', () => {
    const placeholders = [
      {
        raw: '{{ name | pretty }}',
        position: {
          start: 134,
          end: 153
        }
      }
    ]

    expect(() => {
      resolvePlaceholders(placeholders, { name: 'Antoine' }, { strict: true })
    }).toThrow(UnknownFormatterError)
  })

  test('Undefined variable — strict', () => {
    const placeholders = [
      {
        raw: '{{ surname | capitalize }}',
        position: {
          start: 134,
          end: 153
        }
      }
    ]

    expect(() => {
      resolvePlaceholders(placeholders, { name: 'Antoine' }, { strict: true })
    }).toThrow(UndefinedVariableError)
  })
})
