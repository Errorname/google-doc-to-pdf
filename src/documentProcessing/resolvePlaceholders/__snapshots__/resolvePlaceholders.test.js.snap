// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolvePlaceholders commentPlaceholder 1`] = `
Array [
  Object {
    "position": Object {
      "end": 158,
      "start": 134,
    },
    "raw": "{{ # This is a comment }}",
    "type": "comment",
    "value": "This is a comment",
  },
]
`;

exports[`resolvePlaceholders contentPlaceholder - image (width) 1`] = `
Array [
  Object {
    "input": Object {
      "raw": "catUrl",
      "value": "http://example.cat",
    },
    "output": Object {
      "height": undefined,
      "url": "http://example.cat",
      "width": 150,
    },
    "pipes": Array [
      Object {
        "output": Object {
          "height": undefined,
          "url": "http://example.cat",
          "width": 150,
        },
        "raw": "image(150)",
      },
    ],
    "position": Object {
      "end": 166,
      "start": 134,
    },
    "raw": "{{ catUrl | image(150) }}",
    "type": "content",
  },
]
`;

exports[`resolvePlaceholders contentPlaceholder - image (with, height) 1`] = `
Array [
  Object {
    "input": Object {
      "raw": "catUrl",
      "value": "http://example.cat",
    },
    "output": Object {
      "height": 80,
      "url": "http://example.cat",
      "width": 150,
    },
    "pipes": Array [
      Object {
        "output": Object {
          "height": 80,
          "url": "http://example.cat",
          "width": 150,
        },
        "raw": "image(150, 80)",
      },
    ],
    "position": Object {
      "end": 166,
      "start": 134,
    },
    "raw": "{{ catUrl | image(150, 80) }}",
    "type": "content",
  },
]
`;

exports[`resolvePlaceholders contentPlaceholder - image 1`] = `
Array [
  Object {
    "input": Object {
      "raw": "catUrl",
      "value": "http://example.cat",
    },
    "output": Object {
      "height": undefined,
      "url": "http://example.cat",
      "width": undefined,
    },
    "pipes": Array [
      Object {
        "output": Object {
          "height": undefined,
          "url": "http://example.cat",
          "width": undefined,
        },
        "raw": "image",
      },
    ],
    "position": Object {
      "end": 166,
      "start": 134,
    },
    "raw": "{{ catUrl | image }}",
    "type": "content",
  },
]
`;

exports[`resolvePlaceholders contentPlaceholder - text 1`] = `
Array [
  Object {
    "input": Object {
      "raw": "user.name.first",
      "value": "Thibaud",
    },
    "output": "thibaud",
    "pipes": Array [
      Object {
        "output": "thibaud",
        "raw": "lowercase",
      },
    ],
    "position": Object {
      "end": 166,
      "start": 134,
    },
    "raw": "{{ user.name.first | lowercase }}",
    "type": "content",
  },
]
`;

exports[`resolvePlaceholders unknownPlaceholder - reserved symbol 1`] = `
Array [
  Object {
    "position": Object {
      "end": 158,
      "start": 134,
    },
    "raw": "{{ % reserved symbol }}",
  },
]
`;
